lwm2m-client-ipso-digital-input.c

```diff
@@ -41,404 +41,262 @@
+                    {
+                        memcpy(digitalInputs[objectInstanceID].SensorType, *dataPointer, *dataSize);
+                        digitalInputs[objectInstanceID].SensorType[*dataSize] = '\0';
+                    }
+                    else
+                    {
+                        result = AwaResult_BadRequest;
+                    }
+                    break;
+
+                default:
+                    printf("\n Invalid resource ID for DigitalInput write operation");
+                    result = AwaResult_InternalError;
+                    break;
+            }
+            break;
+        default:
+            printf("DigitalInput - unknown operation\n");
+            break;
+    }
+    return result;
 }
 
-/***************************************************************************************************
- * Implementation - Public
- **************************************************************************************************/
-
-int DigitalInput_RegisterDigitalInputObject(Lwm2mContextType * context)
+int DefineDigitalInputObject(AwaStaticClient *awaClient)
 {
-
-	REGISTER_OBJECT(context, "DigitalInput", IPSO_DIGITAL_INPUT_OBJECT, \
-		MultipleInstancesEnum_Multiple, MandatoryEnum_Optional, \
-		&DigitalInputObjectOperationHandlers);
-
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "State", IPSO_DIGITAL_INPUT_STATE, \
-		ResourceTypeEnum_TypeBoolean, Operations_R);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "Counter", IPSO_DIGITAL_INPUT_COUNTER, \
-		ResourceTypeEnum_TypeInteger, Operations_R);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "Polarity", IPSO_DIGITAL_INPUT_POLARITY, \
-		ResourceTypeEnum_TypeBoolean, Operations_RW);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "DebouncePeriod", IPSO_DIGITAL_INPUT_DEBOUNCE_PERIOD, \
-		ResourceTypeEnum_TypeInteger, Operations_RW);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "EdgeSelection", IPSO_DIGITAL_INPUT_EDGE_SELECTION, \
-		ResourceTypeEnum_TypeInteger, Operations_RW);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "CounterReset", IPSO_DIGITAL_INPUT_COUNTER_RESET, \
-		ResourceTypeEnum_TypeNone, Operations_E);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "ApplicationType", IPSO_APPICATION_TYPE, \
-		ResourceTypeEnum_TypeString, Operations_R);
-	REGISTER_DIGITAL_INPUT_RESOURCE(context, "SensorType", IPSO_SENSOR_TYPE, \
-		ResourceTypeEnum_TypeString, Operations_R);
-
-	return 0;
+    AwaError error;
+    error = AwaStaticClient_DefineObjectWithHandler(awaClient, "DigitalInput", IPSO_DIGITAL_INPUT_OBJECT, 0, DIGITAL_INPUTS, digitalInputHandler);
+    if (error != AwaError_Success)
+    {
+        printf("Failed to register flow access object\n");
+        return 1;
+    }
+
+    error = AwaStaticClient_DefineResourceWithHandler(awaClient, "State", IPSO_DIGITAL_INPUT_OBJECT, IPSO_DIGITAL_INPUT_STATE, AwaResourceType_Boolean, 0, 1, AwaResourceOperations_ReadOnly, digitalInputHandler);
```

Comment:
IPSO Smart Objects 1.0 spec says this should be mandatory - change minimumInstances to 1:

    error = AwaStaticClient_DefineResourceWithHandler(awaClient, "State", IPSO_DIGITAL_INPUT_OBJECT, IPSO_DIGITAL_INPUT_STATE, AwaResourceType_Boolean, 1, 1, AwaResourceOperations_ReadOnly, digitalInputHandler);

The digital input state state doesn't toggle with button press but to follow the spec the resource should be created. I'm not sure if there is a way to get down/up events with the button sensor in contiki, but if there is a way this resource should be updated with the button state.